expo/expo
# [expo-updates][android] Add code signing

# Why

This is the android implementation of the code signing portion of #14853.

# How

This introduces two new fields to the update config:
- `expo.modules.updates.CODE_SIGNING_CERTIFICATE` - required config (when code signing is desired) containing an X.509 certificate. The certificate is structured the same as the values set by default when a code signing certificate is generated in Apple Keychain Access, which seems to be pretty standard. To be valid, it must:
     - be valid (dates)
     - have digital signature (key usage)
     - have the code signing extended key usage
- `expo.modules.updates.CODE_SIGNING_METADATA` - optional JSON object containing:
    - `alg` algorithm that should be used to verify the manifest signature. Note that this can differ from the algorithm used to sign the certificate even if the same key is used (rsa-sha256 vs rsa-sha512 for example use the same key but different hashes). This is also used by the client to hint to the server what algorithm it is planning to use to verify the signature.
    - `keyid` an identifier for the key present in the certificate. Defaults to `root` (maybe we should rethink this). This is used by the client to hint to the server what key it is planning to use to verify the signature.

The signature verification process is as follows:
1. Check for certificate. If not present, assume code signing is not enabled and proceed without signature verification.
2. Validate certificate and metadata.
3. Send manifest request with `expo-expects-signature` header following the spec.
3. Upon receiving manifest response (in either json or multipart format), extract the signature.
4. Before constructing the manifest object, verify the signature against the manifest body bytes. If it is valid, continue, else throw.

This also adds an important piece to asset downloading which is to verify that the hash of the downloaded asset matches the expected hash from the manifest. This way the assets are also transitively signed by a signed manifest.

# Test Plan

Run tests.

Also follow the following steps:
1. `yarn link` the local `expo-updates` package
2. Modify `UpdatesPackage.kt` so that linked stuff works.
3. Use in https://github.com/expo/custom-expo-updates-server/pull/2 which uses multipart and code signing.
4. Create and release updates, ensure signing works.
